File: artifacts/snudown/refactor/stack.c
#include <stack.h>
#include <stdlib.h>
#include <string.h>

struct stack
{
  void **item;
  size_t size;
  size_t asize;
};
void stack_free(struct stack *);
int stack_grow(struct stack *, size_t);
int stack_init(struct stack *, size_t);
int stack_push(struct stack *, void *);
void *stack_pop(struct stack *);
void *stack_top(struct stack *);
int stack_grow(struct stack *st, size_t new_size)
{
  void **new_st;
  unsigned int new_st_idx = 0;
  if (st->asize >= new_size)
  {
    return 0;
  }
  new_st_idx = realloc(st->item, new_size * (sizeof(void *)));
  if ((&new_st[new_st_idx]) == 0)
  {
    return -1;
  }
  memset((&new_st[new_st_idx]) + st->asize, 0x0, (new_size - st->asize) * (sizeof(void *)));
  st->item = &new_st[new_st_idx];
  st->asize = new_size;
  if (st->size > new_size)
  {
    st->size = new_size;
  }
  return 0;
}

void stack_free(struct stack *st)
{
  if (!st)
  {
    return;
  }
  free(st->item);
  st->item = 0;
  st->size = 0;
  st->asize = 0;
}

int stack_init(struct stack *st, size_t initial_size)
{
  st->item = 0;
  st->size = 0;
  st->asize = 0;
  if (!initial_size)
  {
    initial_size = 8;
  }
  return stack_grow(st, initial_size);
}

void *stack_pop(struct stack *st)
{
  if (!st->size)
  {
    return 0;
  }
  return st->item[--st->size];
}

int stack_push(struct stack *st, void *item)
{
  if (stack_grow(st, st->size * 2) < 0)
  {
    return -1;
  }
  st->item[st->size++] = item;
  return 0;
}

void *stack_top(struct stack *st)
{
  if (!st->size)
  {
    return 0;
  }
  return st->item[st->size - 1];
}

File: artifacts/snudown/translate_ablation/src/stack.rs
use crate::*;
use std::ptr;
pub struct Stack {
    pub item: Option<Vec<Option<Box<dyn std::any::Any>>>>,
    pub size: usize,
    pub asize: usize,
}

pub fn stack_grow(st: &mut Stack, new_size: usize) -> i32 {
    if st.asize >= new_size {
        return 0;
    }
    
    // Get the current vector or create a new one if None
    let mut vec = st.item.take().unwrap_or_else(|| Vec::new());
    
    // Resize the vector to the new capacity, filling with None
    vec.resize_with(new_size, || None);
    
    // Update the stack with the new vector
    st.item = Some(vec);
    st.asize = new_size;
    
    if st.size > new_size {
        st.size = new_size;
    }
    
    0
}
pub fn stack_push(st: &mut Stack, item: Option<Box<dyn std::any::Any>>) -> i32 {
    if stack_grow(st, st.size * 2) < 0 {
        return -1;
    }
    if let Some(ref mut items) = st.item {
        items.push(item);
        st.size += 1;
        0
    } else {
        -1
    }
}
pub fn stack_free(st: &mut Stack) {
    if st.item.is_none() {
        return;
    }
    
    // Free the Vec by replacing it with None
    st.item = None;
    st.size = 0;
    st.asize = 0;
}
pub fn stack_pop(st: &mut Stack) -> Option<Box<dyn std::any::Any>> {
    if st.size == 0 {
        return None;
    }
    st.size -= 1;
    st.item.as_mut()?.pop().flatten()
}
pub fn stack_top(st: &Stack) -> Option<&Box<dyn std::any::Any>> {
    if st.size == 0 {
        return None;
    }
    st.item.as_ref().and_then(|v| v[st.size - 1].as_ref())
}
pub fn stack_init(st: &mut Stack, initial_size: usize) -> i32 {
    st.item = None;
    st.size = 0;
    st.asize = 0;
    let size = if initial_size == 0 { 8 } else { initial_size };
    stack_grow(st, size)
}
